lib/webidl2.js

```diff
@@ -15,57 +15,62 @@
     "other": /[^\t\n\r 0-9A-Z_a-z]/y
   };
 
-  function attemptTokenMatch(str, type, lastIndex, tokens) {
-    const re = tokenRe[type];
-    re.lastIndex = lastIndex;
-    const result = re.exec(str);
-    if (result) {
-      tokens.push({ type, value: result[0] });
-      return re.lastIndex;
-    }
-    return -1;
-  }
-
   function tokenise(str) {
     const tokens = [];
     let lastIndex = 0;
+    let trivia = "";
     while (lastIndex < str.length) {
       const nextChar = str.charAt(lastIndex);
       let result = -1;
-      if (/[-0-9.]/.test(nextChar)) {
-        result = attemptTokenMatch(str, "float", lastIndex, tokens);
+
+      if (/[\t\n\r ]/.test(nextChar)) {
+        result = attemptTokenMatch("whitespace", true);
+      } else if (nextChar === '/') {
+        result = attemptTokenMatch("comment", true);
+      }
+
+      if (result !== -1) {
+        trivia += tokens.pop().value;
+      } else if (/[-0-9.]/.test(nextChar)) {
+        result = attemptTokenMatch("float");
         if (result === -1) {
-          result = attemptTokenMatch(str, "integer", lastIndex, tokens);
+          result = attemptTokenMatch("integer");
         }
         if (result === -1) {
           // '-' and '.' can also match "other".
-          result = attemptTokenMatch(str, "other", lastIndex, tokens);
+          result = attemptTokenMatch("other");
         }
       } else if (/[A-Z_a-z]/.test(nextChar)) {
-        result = attemptTokenMatch(str, "identifier", lastIndex, tokens);
+        result = attemptTokenMatch("identifier");
       } else if (nextChar === '"') {
-        result = attemptTokenMatch(str, "string", lastIndex, tokens);
+        result = attemptTokenMatch("string");
         if (result === -1) {
           // '"' can also match "other".
-          result = attemptTokenMatch(str, "other", lastIndex, tokens);
-        }
-      } else if (/[\t\n\r ]/.test(nextChar)) {
-        result = attemptTokenMatch(str, "whitespace", lastIndex, tokens);
-      } else if (nextChar === '/') {
-        result = attemptTokenMatch(str, "comment", lastIndex, tokens);
-        if (result === -1) {
-          // '/' can also match "other".
-          result = attemptTokenMatch(str, "other", lastIndex, tokens);
+          result = attemptTokenMatch("other");
         }
       } else {
-        result = attemptTokenMatch(str, "other", lastIndex, tokens);
+        result = attemptTokenMatch("other");
       }
       if (result === -1) {
         throw new Error("Token stream not progressing");
       }
       lastIndex = result;
     }
     return tokens;
+
+    function attemptTokenMatch(type, noFlushTrivia) {
```

Comment:
Nit: we should try to avoid introducing another boolean trap. Can you make the second argument an object?